### PR TITLE
`Adaptive learning`: Improve learning path sidebar in participation view

### LINKED ISSUE #7176: `Adaptive learning`: Make learning path graph in participation view scrollable.
### Is your feature request related to a problem?

_No response_

### Describe the solution you'd like

The learning path graph in the participation view should be scrollable.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by MaximilianAnzinger**
⚠️ Stacked on #7148  please don't merge accidentally
======

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

In order to make the UI in the participation view more intuitive, we want to replace the sidebar with a component that is tailored to the specific needs of a strictly linear representation of a learning path.

Resolves #7176 

### Description

- replaces the sidebar component with an improved implementation
- removed view encapsulation none from LearningPathGraphComponent

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Course with a learning path enabled and at least one competency with many exercises

1. Log in to Artemis
2. Navigate to the Learning Paths tab in the course
3. check that the left sidebar is scrollable
4. check that the left sidebar is sticky (it does not disappear if you scroll down on a long exercise)
5. press the two up and down buttons, make sure the current lecture unit or exercise is highlighted and scrolled into view
6. press on the button on top of the sidebar
7. make sure the complete learning path graph opens and is displayed as expected

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [learning-path-graph.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5459/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/learning-paths/learning-path-graph/learning-path-graph.component.ts.html) | 87.8% | ✅  |
| [learning-path.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5459/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/learning-paths/learning-path-graph/learning-path.component.ts.html) | 97.14% | ✅  |
| [learning-path-container.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5459/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/learning-paths/participate/learning-path-container.component.ts.html) | 84.69% | ✅  |
| [learning-path-storage.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5459/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/learning-paths/participate/learning-path-storage.service.ts.html) | 96.29% | ✅  |
| [learning-path-progress-modal.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5459/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/learning-paths/progress-modal/learning-path-progress-modal.component.ts.html) | 100% | ✅ |
| [learning-path.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5459/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/competency/learning-path.model.ts.html) | 95.45% | ✅  |

### Screenshots

Screenshot 2023-09-27 at 16 14 57

### CODE REVIEW

**Review by florian-glombik — APPROVED** [review]
Tested on ts2, the key functionality of this PR works fine, I do only have some comments on how the UI could be used differently.

* For me it is uncomfortable to click the scrollbar, I would prefer to use the scroll of my mouse wheel when hovering over the graph
image
* I would use the arrows to navigate in the navigation bar - it is unintuitive for me that I cannot use the arrow to scroll up to the goal
image
* a general problem in Artemis, however it would be great if we would start fixing the breadcrumbs - if navigating to an exercise from the learning path the breadcrumb does not lead back to the learning path
image

**Review by YeetTheFirst21 — APPROVED** [review]
Tested on TS2 in the testing session,
aggree with the changes others said but also believe that both the scrollbar and the buttons should stay.

Overall a nice adition that can get improved over time

The buttons can possibly also go thrugh the competency headings as well as the exercises.

The eye icon can also possibly go over the up button, possibly like this:

Screenshot 2023-09-21 142903

File: src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path.component.scss [hunk 1]
```diff
@@ -0,0 +1,31 @@
+.path-wrapper {
```
Comment (JohannesStoehr) [inline src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path.component.scss:1]: Is this empty on purpose?

File: src/main/webapp/app/entities/competency/learning-path.model.ts [hunk 2]
```diff
@@ -34,6 +35,30 @@ export class NgxLearningPathNode implements Node {
...
+    // return different icons for completed learning objects
+    if (node.type === NodeType.EXERCISE || node.type === NodeType.LECTURE_UNIT) {
+        if (node.completed) {
+            return faCheckCircle;
+        } else {
+            return faPlayCircle;
+        }
+    }
+
+    const icons = {
```
Comment (JohannesStoehr) [inline src/main/webapp/app/entities/competency/learning-path.model.ts:53]: Why not use a switch here?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/entities/competency/learning-path.model.ts (hunk 2 above)
Comment (MaximilianAnzinger) [reply src/main/webapp/app/entities/competency/learning-path.model.ts:53]: This follows the pattern from other getIcon Methods (lecture units, exercises, competencies). Also, with formatting a switch takes up double the amount of lines and makes it more difficult to read imo.

Comment (JohannesStoehr) [reply src/main/webapp/app/entities/competency/learning-path.model.ts:53]: Why doesn't JavaScript have a switch expression that you can immediately return 😢
Then it's fine to me

**Review by JohannesStoehr — DISMISSED** [review]
Code looks good to me now

**Review by lennart-keller — DISMISSED** [review]
Code looks good to me

**Review by rriyaldhi — APPROVED** [review]
Tested on TS2 and works as described on the testing steps.

**Review by pal03377 — APPROVED** [review]
tested in testing session on ts2 without major issues

**Review by tobias-lippert — DISMISSED** [review]
Everything worked. Nice UI improvement.

**Review by rstief — DISMISSED** [review]
Tested on TS2

**Review by YeetTheFirst21 — APPROVED** [review]
Looks really good with simpler button navigation

File: src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path.component.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    templateUrl: './learning-path.component.html',
+    encapsulation: ViewEncapsulation.None,
+})
+export class LearningPathComponent implements OnInit {
+    @Input() learningPathId: number;
+    @Input() courseId: number;
+    @Output() nodeClicked: EventEmitter<NgxLearningPathNode> = new EventEmitter();
+
+    isLoading = false;
+    path?: NgxLearningPathNode[];
```
Comment (Strohgelaender) [inline src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path.component.ts:19]:
```suggestion
    path: NgxLearningPathNode[] = [];
```

File: src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path.component.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        this.loadData();
+    }
+
+    private loadData() {
+        if (!this.learningPathId) {
+            return;
+        }
+        this.isLoading = true;
+        this.learningPathService.getLearningPathNgxPath(this.learningPathId).subscribe((ngxLearningPathResponse) => {
+            this.path = [];
```
Comment (Strohgelaender) [inline src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path.component.ts:42]: no longer needed if the default value above is used.
```suggestion
```

**Review by Strohgelaender — DISMISSED** [review]
code

**Review by JohannesStoehr — DISMISSED** [review]
Reapprove code after merges and base branch change

**Review by rstief — DISMISSED** [review]
reapproved.

**Review by tobias-lippert — APPROVED** [review]
Reapprove

**Review by Strohgelaender — DISMISSED** [review]
re-approve

File: src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path.component.scss [hunk 5]
```diff
@@ -0,0 +1,29 @@
+.path-edge {
```
[the remark this answers is not in the document]
Comment (MaximilianAnzinger) [reply src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path.component.scss:1]: I've managed to enable the encapsulation again. Thx for pointing this out.

File: src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path-graph.component.scss [hunk 6]
```diff
@@ -41,15 +41,15 @@ jhi-learning-path-graph-node:hover {
...
-
-        svg {
-            margin: 10%;
-            width: 80%;
-            height: 80%;
-        }
     }
 }
 
+::ng-deep .graph-container .node-icon-container fa-icon svg {
```
[the remark this answers is not in the document]
Comment (MaximilianAnzinger) [reply src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path-graph.component.scss]: For now, this is the best solution I could come up with. I could not get the styling to work without causing possible side effects.

**Review by Strohgelaender — APPROVED** [review]

**Review by rstief — APPROVED** [review]

**Review by lennart-keller — APPROVED** [review]
reapprove, code still lgtm

**Review by tobias-lippert — APPROVED** [review]
Reapprove

**Pull request merged**